ggml/include/ggml.h

```diff
@@ -624,13 +626,26 @@ extern "C" {
     // If it returns true, the computation is aborted
     typedef bool (*ggml_abort_callback)(void * data);
 
+    struct ggml_threadpool_params {
+        bool     cpumask[GGML_MAX_N_THREADS]; // mask of cpu cores
+        bool     mask_specified;              // mask is non-empty
+        int32_t  n_threads;                   // number of threads
+        int32_t  prio;                        // thread priority
+        uint32_t poll;                        // polling level (0 - no polling, 100 - aggressive polling)
+        bool     strict_cpu;                  // strict cpu placement
+        bool     paused;                      // start in paused state
+    };
```

Comment:
Yep. `mask_specified` can go, the default for the cpumask used to all-ones so there was no way to tell if it was set. We can just make the default all-zeros and not change the affinity in that case. Will update tomorrow.

I'd rather keep the `poll` exposed. It depends on the latency requirements (and useful for benchmarking). The default might be OK now but might be off for newer/unusual machines. No harm in letting user tweek it (and perhaps suggest updated default).

`paused` is needed when we have two threadpools (one for batch and one for non-batch). If we don't start the second pool in paused state there is a regression in batch processing where both pools are running initially (especially if aggressive polling is enabled).

Reply:
I am not opposed to allowing users to change the value of `poll`, but if it is done here, it forces the users to set a value without a default. Instead, there should be a good default, and users should only need to think about this if they want to change it.

The regression with `pause` could be solved by always starting the pools from the sleep state. I don't think there is a good reason to start spinning as soon as the pool is created, except maybe for the temporary pool that is created when there is no pool attached, but that's an implementation detail that doesn't need to be exposed in the public API.

Reply:
Always starting paused (aka sleeping, which was the original default) also causes regressions. If you follow my commits all these little tricks were added as follow up to the original regressions (some of which you reported a couple of weeks back). For the small models and/or short prompts if we start sleeping we end up having to go through two context switches (in each thread) and a wakeup in between (especially noticeable on larger number of threads say 32+ on EPYC). The threads start (ie from `pthread_create()`) and immediately go to sleep, even though we have immediate work for them, we then wake them up and eat that wake up latency as part of prompt processing. For larger models it's not too big of an impact but for smaller models (or very short prompts of a few tokens) it shows up right away.
Also exposing this in the API is OK IMHO. It just gives uses ability to create multiple pool in paused states to reuse in the future. As you said, as long as there is a good default users should not need to care about it.

So all these has been carefully tested and profiled, that's why we're seeing good perf in most cases (looks like besides that regression on your M3).

Now, I totally agree with the need for a good default for `poll` (and all other values) in the public API though. 
If this struct was C++ we would've added a nice default there. What's you preference for C API in this case? 
We could add `void ggml_threadpool_params_init_defaults(struct threadpool_params *p);`.

Reply:
Adding a few microseconds to wake up the threads in the first run shouldn't affect performance significantly. `llama-bench` discards the first run, and subsequent runs should be identical regardless of the initial pause state of the threads. Other examples that use the common library to initialize llama.cpp also run and discard a warmup run first, so it shouldn't show there either. So I am not sure why starting the threads paused would cause a performance regression in the tests.

The way I look at this is that we should avoid exposing details in the API unless there is a clear need to have them, since it increases complexity and it increases the chances of breaking user code when making changes to the library.

A function to init the params could work. `struct threadpool_params ggml_threadpool_default_params(void)` would be more in line with other similar functions already present in the API.

Reply:
Unfortunately, it's not a few usec. As I mentioned, it's two context switches for each thread, sleep/wakeup in the middle. 
I'll share the benchmarks with/without in a bit. 

I added support for the `ggml_threadpool_params_default/init()`. Testing it now on all platforms. 
Will push in a bit and ping you to take a look. That `paused` param is now really hidden and doesn't 
add any extra logic. It's literally just the init value for the default worker state, that's it.
4 lines of code (5 if you count the comment :) ) that cleanly address the corner case I described.
```
~/src/llama.cpp-threadpool$ git grep paused
examples/main/main.cpp:        // Start the non-batch threadpool in the paused state
examples/main/main.cpp:        tpp.paused = true;
ggml/include/ggml.h:        bool     paused;                      // start in paused state
ggml/src/ggml.c:    p->paused     = false; // threads are ready to go
ggml/src/ggml.c:        threadpool->pause            = tpp->paused;
```

Looks pretty clean to me, again push is coming shortly once I double check I didn't break anything on Ryzen, M2, X-Elite, ...

Reply:
@slaren 

Here are some number and an easy way to demonstrate the benefit. I built the latest threadpool branch with `threadpool.paused=true` as the default and compared against `threadpool.paused=false`.

As the number of threads grows the context switches and wakeup latency adds up.
See two simple use-cases below: short prompt with `llama-cli` and stories260K with `llama-bench` on Ryzen9.
With 4 threads there is not much difference, but with 16 threads PP numbers are quite different.

Also see the number of context switches for the bench runs. The runs are very short but we consistently see almost 2x number of context switches.   

<details>
AMD Ryzen 9 3950x

```
llama.cpp-threadpool$ ./llama-cli-pause  -m ../gguf/llama-v3.1-8b-instruct.q4_k_m.gguf -ngl 0 --ctx-size 4096 -t 4 --seed 42 -p 'what is the most popular cookie in the world now? (please be brief)' -n 40 |& grep timings
llama_print_timings:        load time =     804.39 ms
llama_print_timings:      sample time =       2.91 ms /    40 runs   (    0.07 ms per token, 13726.84 tokens per second)
llama_print_timings: prompt eval time =     852.80 ms /    16 tokens (   53.30 ms per token,    18.76 tokens per second)
llama_print_timings:        eval time =    4293.30 ms /    39 runs   (  110.08 ms per token,     9.08 tokens per second)
llama_print_timings:       total time =    5159.20 ms /    55 tokens

llama.cpp-threadpool$ ./llama-cli-nopause  -m ../gguf/llama-v3.1-8b-instruct.q4_k_m.gguf -ngl 0 --ctx-size 4096 -t 4 --seed 42 -p 'what is the most popular cookie in the world now? (please be brief)' -n 40 |& grep timings
llama_print_timings:        load time =     805.53 ms
llama_print_timings:      sample time =       2.90 ms /    40 runs   (    0.07 ms per token, 13778.85 tokens per second)
llama_print_timings: prompt eval time =     850.83 ms /    16 tokens (   53.18 ms per token,    18.81 tokens per second)
llama_print_timings:        eval time =    4298.50 ms /    39 runs   (  110.22 ms per token,     9.07 tokens per second)
llama_print_timings:       total time =    5162.84 ms /    55 tokens

llama.cpp-threadpool$ ./llama-cli-pause  -m ../gguf/llama-v3.1-8b-instruct.q4_k_m.gguf -ngl 0 --ctx-size 4096 -t 16 --seed 42 -p 'what is the most popular cookie in the world now? (please be brief)' -n 40 |& grep timings
llama_print_timings:        load time =     763.45 ms
llama_print_timings:      sample time =       2.96 ms /    40 runs   (    0.07 ms per token, 13508.95 tokens per second)
llama_print_timings: prompt eval time =     300.52 ms /    16 tokens (   18.78 ms per token,    53.24 tokens per second) <<<<<
llama_print_timings:        eval time =    4192.73 ms /    39 runs   (  107.51 ms per token,     9.30 tokens per second)
llama_print_timings:       total time =    4502.54 ms /    55 tokens

llama.cpp-threadpool$ ./llama-cli-nopause  -m ../gguf/llama-v3.1-8b-instruct.q4_k_m.gguf -ngl 0 --ctx-size 4096 -t 16 --seed 42 -p 'what is the most popular cookie in the world now? (please be brief)' -n 40 |& grep timings
llama_print_timings:        load time =     763.25 ms
llama_print_timings:      sample time =       2.94 ms /    40 runs   (    0.07 ms per token, 13605.44 tokens per second)
llama_print_timings: prompt eval time =     261.23 ms /    16 tokens (   16.33 ms per token,    61.25 tokens per second) <<<<<
llama_print_timings:        eval time =    4176.01 ms /    39 runs   (  107.08 ms per token,     9.34 tokens per second)
llama_print_timings:       total time =    4446.36 ms /    55 tokens
```



$ perf stat ./llama-bench-pause -m ../gguf/stories260K.gguf -ngl 0 -r 10 -t 32 -p 512 -n 0
| model                          |       size |     params | backend    | threads |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | ------------: | ---------------: |
| llama ?B all F32 (guessed)     |   1.12 MiB |     0.29 M | CPU        |      32 |         pp512 | 38068.49 ± 4766.49 |

build: 76d2c517 (3640)

 Performance counter stats for './llama-bench-pause -m ../gguf/stories260K.gguf -ngl 0 -r 10 -t 32 -p 512 -n 0':

          7,399.38 msec task-clock                       #   30.002 CPUs utilized             
               171      context-switches                 #   23.110 /sec                      
                33      cpu-migrations                   #    4.460 /sec                      
            65,035      page-faults                      #    8.789 K/sec                     
    29,704,525,025      cycles                           #    4.014 GHz                         (83.38%)
     4,820,244,067      stalled-cycles-frontend          #   16.23% frontend cycles idle        (83.09%)
     2,904,276,853      stalled-cycles-backend           #    9.78% backend cycles idle         (83.09%)
    18,418,923,615      instructions                     #    0.62  insn per cycle            
                                                  #    0.26  stalled cycles per insn     (83.18%)
     2,553,065,215      branches                         #  345.038 M/sec                       (83.48%)
        18,433,130      branch-misses                    #    0.72% of all branches             (84.36%)

       0.246633517 seconds time elapsed

       6.226568000 seconds user
       1.121580000 seconds sys


$ perf stat ./llama-bench-nopause -m ../gguf/stories260K.gguf -ngl 0 -r 10 -t 32 -p 512 -n 0
| model                          |       size |     params | backend    | threads |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | ------------: | ---------------: |
| llama ?B all F32 (guessed)     |   1.12 MiB |     0.29 M | CPU        |      32 |         pp512 | 40999.11 ± 209.90 |

build: 76d2c517 (3640)

 Performance counter stats for './llama-bench-nopause -m ../gguf/stories260K.gguf -ngl 0 -r 10 -t 32 -p 512 -n 0':

          5,691.93 msec task-clock                       #   30.604 CPUs utilized             
               103      context-switches                 #   18.096 /sec                      
                 0      cpu-migrations                   #    0.000 /sec                      
            64,973      page-faults                      #   11.415 K/sec                     
    23,885,300,316      cycles                           #    4.196 GHz                         (83.96%)
     4,872,936,678      stalled-cycles-frontend          #   20.40% frontend cycles idle        (82.02%)
     2,222,070,039      stalled-cycles-backend           #    9.30% backend cycles idle         (82.13%)
    18,123,432,355      instructions                     #    0.76  insn per cycle            
                                                  #    0.27  stalled cycles per insn     (83.54%)
     2,566,348,648      branches                         #  450.875 M/sec                       (84.26%)
        18,727,582      branch-misses                    #    0.73% of all branches             (84.22%)

       0.185984183 seconds time elapsed

       4.448328000 seconds user
       1.235205000 seconds sys

</details>

Reply:
A possible solution could be to delay creation of the threads until the first time the pool is used in `ggml_graph_compute`.

Reply:
> A possible solution could be to delay creation of the threads until the first time the pool is used in `ggml_graph_compute`.

That adds more special cases. I mentioned earlier (in commit comments, etc) that I went through and reduced special casing as much as possible to avoid branches and simplify things in general. Being able to init the state of the threadpool (ie start paused) seems like the simplest solution.

Reply:
Right, it would add more complexity to the ggml code, but that's an acceptable tradeoff for simplifying the user code. It is important to remember that ggml is not just a component of llama.cpp, it is a library used in many projects.